File: packages/pvm-assemblyscript/assembly/mock-ram.ts
/**
 * Mock RAM implementation for testing
 * 
 * A no-op implementation that satisfies the RAM interface but does nothing.
 * Useful for testing when you don't need actual memory functionality,
 * or to isolate issues that might be in RAM implementation code.
 */

import { FaultCheckResult, ReadResult, RAM, MemoryAccessType, WriteResult } from './types'
import { MEMORY_CONFIG } from './config'

export class MockRAM implements RAM {
  currentHeapPointer: u32 = 0

  // JIP-6 trace support: Track last load/store (no-op, always 0)
  lastLoadAddress: u32 = 0
  lastLoadValue: u64 = 0
  lastStoreAddress: u32 = 0
  lastStoreValue: u64 = 0

  /**
   * Clear last load/store tracking (no-op)
   */
  clearLastMemoryOp(): void {
    this.lastLoadAddress = 0
    this.lastLoadValue = 0
    this.lastStoreAddress = 0
    this.lastStoreValue = 0
  }

  /**
   * Read multiple octets from memory (no-op)
   * Returns zero-filled data with no faults
   */
  readOctets(address: u32, count: u32): ReadResult {
    // Return zero-filled data, no fault
    return new ReadResult(new Uint8Array(count), 0)
  }

  /**
   * Write multiple octets to memory (no-op)
   * Always succeeds
   */
  writeOctets(address: u32, values: Uint8Array): WriteResult {
    // Always succeed (no fault)
    return new WriteResult(false, 0)
  }

  /**
   * Allocate pages (no-op)
   */
  allocatePages(startPage: u32, count: u32): void {
    // Do nothing
  }

  /**
   * Check if memory is readable (no-op)
   * Always returns readable
   */
  isReadableWithFault(address: u32, size: u32 = u32(1)): FaultCheckResult {
    // Always readable (no fault)
    return new FaultCheckResult(true, 0)
  }

  /**
   * Check if memory is writable (no-op)
   * Always returns writable
   */
  isWritableWithFault(address: u32, size: u32 = u32(1)): FaultCheckResult {
    // Always writable (no fault)
    return new FaultCheckResult(true, 0)
  }

  /**
   * Initialize memory layout (no-op)
   */
  initializeMemoryLayout(
    argumentData: Uint8Array,
    readOnlyData: Uint8Array,
    readWriteData: Uint8Array,
    stackSize: u32,
    heapZeroPaddingSize: u32,
  ): void {
    // Do nothing
  }

  /**
   * Set page access rights (no-op)
   */
  setPageAccessRights(address: u32, length: u32, accessType: MemoryAccessType): void {
    // Do nothing
  }

  /**
   * Initialize a memory page (no-op)
   */
  initPage(address: u32, length: u32, accessType: MemoryAccessType): void {
    // Do nothing
  }

  /**
   * Write to memory during initialization (no-op)
   */
  writeOctetsDuringInitialization(address: u32, values: Uint8Array): void {
    // Do nothing
  }

  /**
   * Get page dump for a specific page index (no-op)
   * Returns zeros for mock implementation
   */
  getPageDump(pageIndex: u32): Uint8Array {
    // Return zeros (mock implementation)
    return new Uint8Array(MEMORY_CONFIG.PAGE_SIZE)
  }

  /**
   * Reset RAM to initial state
   * Resets heap pointer (no-op for mock)
   */
  reset(): void {
    // Reset heap pointer
    this.currentHeapPointer = 0
  }
}

File: packages/pvm-rust/src/mock_ram.rs
//! Mock RAM for tests (mirrors assembly/mock-ram.ts).
//! No-op implementation that satisfies the RAM interface.

use crate::config;
use crate::types::{FaultCheckResult, MemoryAccessType, Ram, ReadResult, WriteResult};

/// Mock RAM: no-op memory, always succeeds / returns zeros.
#[derive(Default)]
pub struct MockRam {
    current_heap_pointer: u32,
    last_load_address: u32,
    last_load_value: u64,
    last_store_address: u32,
    last_store_value: u64,
}

impl MockRam {
    #[must_use]
    pub const fn new() -> Self {
        Self {
            current_heap_pointer: 0,
            last_load_address: 0,
            last_load_value: 0,
            last_store_address: 0,
            last_store_value: 0,
        }
    }
}

impl Ram for MockRam {
    fn read_octets(&mut self, _address: u32, count: u32) -> ReadResult {
        ReadResult::new(Some(vec![0u8; count as usize]), 0)
    }

    fn write_octets(&mut self, _address: u32, _values: &[u8]) -> WriteResult {
        WriteResult::new(false, 0)
    }

    fn current_heap_pointer(&self) -> u32 {
        self.current_heap_pointer
    }

    fn set_current_heap_pointer(&mut self, value: u32) {
        self.current_heap_pointer = value;
    }

    fn allocate_pages(&mut self, _start_page: u32, _count: u32) {}

    fn is_readable_with_fault(&self, _address: u32, _size: u32) -> FaultCheckResult {
        FaultCheckResult::new(true, 0)
    }

    fn initialize_memory_layout(
        &mut self,
        _argument_data: &[u8],
        _read_only_data: &[u8],
        _read_write_data: &[u8],
        _stack_size: u32,
        _heap_zero_padding_size: u32,
    ) {
    }

    fn is_writable_with_fault(&self, _address: u32, _size: u32) -> FaultCheckResult {
        FaultCheckResult::new(true, 0)
    }

    fn set_page_access_rights(&mut self, _address: u32, _length: u32, _access_type: MemoryAccessType) {
    }

    fn init_page(&mut self, _address: u32, _length: u32, _access_type: MemoryAccessType) {}

    fn write_octets_during_initialization(&mut self, _address: u32, _values: &[u8]) {}

    fn get_page_dump(&self, _page_index: u32) -> Vec<u8> {
        vec![0u8; config::PAGE_SIZE as usize]
    }

    fn reset(&mut self) {
        self.current_heap_pointer = 0;
        self.last_load_address = 0;
        self.last_load_value = 0;
        self.last_store_address = 0;
        self.last_store_value = 0;
    }

    fn last_load_address(&self) -> u32 {
        self.last_load_address
    }

    fn last_load_value(&self) -> u64 {
        self.last_load_value
    }

    fn last_store_address(&self) -> u32 {
        self.last_store_address
    }

    fn last_store_value(&self) -> u64 {
        self.last_store_value
    }

    fn clear_last_memory_op(&mut self) {
        self.last_load_address = 0;
        self.last_load_value = 0;
        self.last_store_address = 0;
        self.last_store_value = 0;
    }
}
